src/graph/sampler.cc

```diff
@@ -988,415 +988,769 @@ std::vector<dgl_id_t> Global2Local(const std::vector<size_t> &ids,
-  // is 0.
-  int64_t num_chunks = num_pos_edges / chunk_size;
-  int64_t last_chunk_size = num_pos_edges - num_chunks * chunk_size;
-
-  // The number of negative edges.
-  int64_t num_neg_edges = neg_sample_size * chunk_size * num_chunks;
-  int64_t num_neg_edges_last_chunk = neg_sample_size * last_chunk_size;
-  int64_t num_all_neg_edges = num_neg_edges + num_neg_edges_last_chunk;
-
-  // We should include the last chunk.
-  if (last_chunk_size > 0)
-    num_chunks++;
-
-  IdArray neg_dst = IdArray::Empty({num_all_neg_edges}, coo->dtype, coo->ctx);
-  IdArray neg_src = IdArray::Empty({num_all_neg_edges}, coo->dtype, coo->ctx);
-  IdArray neg_eid = IdArray::Empty({num_all_neg_edges}, coo->dtype, coo->ctx);
-  IdArray induced_neg_eid = IdArray::Empty({num_all_neg_edges}, coo->dtype, coo->ctx);
-
-  // These are vids in the positive subgraph.
-  const dgl_id_t *dst_data = static_cast<const dgl_id_t *>(coo->data);
-  const dgl_id_t *src_data = static_cast<const dgl_id_t *>(coo->data) + num_pos_edges;
-  const dgl_id_t *induced_vid_data = static_cast<const dgl_id_t *>(pos_subg.induced_vertices->data);
-  const dgl_id_t *induced_eid_data = static_cast<const dgl_id_t *>(pos_subg.induced_edges->data);
-  size_t num_pos_nodes = pos_subg.graph->NumVertices();
-  std::vector<size_t> pos_nodes(induced_vid_data, induced_vid_data + num_pos_nodes);
 
-  dgl_id_t *neg_dst_data = static_cast<dgl_id_t *>(neg_dst->data);
-  dgl_id_t *neg_src_data = static_cast<dgl_id_t *>(neg_src->data);
-  dgl_id_t *neg_eid_data = static_cast<dgl_id_t *>(neg_eid->data);
-  dgl_id_t *induced_neg_eid_data = static_cast<dgl_id_t *>(induced_neg_eid->data);
+  GraphPtr gptr_;
+  IdArray seed_edges_;
+  IdArray relations_;
 
-  const dgl_id_t *unchanged;
-  dgl_id_t *neg_unchanged;
-  dgl_id_t *neg_changed;
+  int64_t batch_size_;
+  int64_t num_workers_;
+  std::string neg_mode_;
+  int64_t neg_sample_size_;
+  bool exclude_positive_;
+  bool check_false_neg_;
+};
 
-  // corrupt head nodes.
-  if (IsNegativeHeadMode(neg_mode)) {
-    unchanged = dst_data;
-    neg_unchanged = neg_dst_data;
-    neg_changed = neg_src_data;
-  } else {
-    // corrupt tail nodes.
-    unchanged = src_data;
-    neg_unchanged = neg_src_data;
-    neg_changed = neg_dst_data;
+class UniformEdgeSamplerObject: public EdgeSamplerObject {
+public:
+  explicit UniformEdgeSamplerObject(const GraphPtr gptr,
+                                    IdArray seed_edges,
+                                    const int64_t batch_size,
+                                    const int64_t num_workers,
+                                    const std::string neg_mode,
+                                    const int64_t neg_sample_size,
+                                    const bool exclude_positive,
+                                    const bool check_false_neg,
+                                    IdArray relations)
+                                    : EdgeSamplerObject{gptr,
```

Comment:
what is the difference between '{' and '(' here?

Reply:
I think no much difference. Both are OK.

Reply:
can you use '('? '{' looks weird.